lib/msal-common/docs/authority.md

```diff
@@ -0,0 +1,111 @@
+
+```console
+    <scheme>://<domain-of-the-azure-instance>/<tenant-identifier>
+```
+
+> :warning: The canonical authority URI differs in the case of B2C and ADFS. See [Choosing an authority](#choosing-an-authority) below for more.
+
+The authority URI guides MSAL where to look for the 3 endpoints that are required for successfully acquiring a token:
+
+| Endpoint                | Example     | Description |
+|-------------------------|-------------|-------------|
+| `/openid-configuration` | `https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration` | Contains the metadata required for making a token request e.g. supported scopes, claims, request methods, signing keys etc. |
+| `/authorize`            | `https://login.microsoftonline.com/common/oauth2/v2.0/authorize` | Returns an authorization code in the query parameters via the client app's registered redirect URI. Navigation to this endpoint usually prompts the user with a dialog screen by the IdP. Appropriate query parameters can be used to customize this screen. |
+| `/token`                | `https://login.microsoftonline.com/common/oauth2/v2.0/token` | The client app sends the authorizaiton code to this endpoint via a POST request and receives an access and an ID token in return. |
+
+> :bulb: Certain OAuth 2.0 grants may skip the authorize endpoint and go directly for the token endpoint, e.g. [OAuth 2.0 Client Credentials Grant](https://docs.microsoft.com/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow)
+
+## Authority configuration
+
+In MSAL, authority can be set in 2 locations:
+
+1. via the [configuration](https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_common.html#clientconfiguration) object during MSAL instantiation
+2. via the [request](https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_common.html#baseauthrequest) object during `login*` and `acquireToken*` calls
+
+(2) takes precedence over (1) i.e. you can override the initial authority URI set in the configuration object later on during each token request. See [msal-browser configuration](../../msal-browser/docs/configuration.md) and [msal-node configuration](../../msal-node/docs/configuration.md) for more.
+
+## Choosing an authority
+
+The correct authority URI that you need pass to MSAL is ultimately determined by the domain of the IdP you register your application with. When registering an application, consider:
+
+- Who are the users you want your app to sign-in?
+- What are the web APIs you want your app to acquire tokens for?
+
+### Azure AD
+
+The authority domain for the global Azure AD instance is `login.microsoftonline.com`. This domain has several aliases (e.g. `login.microsoft.com`), and MSAL keeps a record of these. MSAL trusts authority URIs with any of these aliases by default.
```

Comment:
Should we add a link to the [`discovery endpoint`](https://login.microsoftonline.com/common/discovery/instance?api-version=1.1&authorization_endpoint=https%3A%2F%2Flogin.microsoftonline.com%2Fcommon%2Foauth2%2Fv2.0%2Fauthorize) where these aliases are derived from? Also @samuelkubai did some work to cache these locally etc. Links to them would be helpful.

Reply:
Sure thing, added a remark